lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -1357,6 +1402,285 @@ class DifferentiableActivityInfo {
+  // found extracted element is useful.
+  // Collect direct results being retrieved using `tuple_extract`.
+  SmallVector<SILValue, 8> usedDirectResults(convs.getNumDirectSILResults());
+  for (auto *use : ai->getUses())
+    if (auto *tei = dyn_cast<TupleExtractInst>(use->getUser()))
+      usedDirectResults[tei->getFieldNo()] = tei;
+  // Add differentiation indices based on activity analysis.
+  unsigned dirResIdx = 0;
+  unsigned indResIdx = convs.getSILArgIndexOfFirstIndirectResult();
+  for (auto &resAndIdx : enumerate(convs.getResults())) {
+    auto &res = resAndIdx.value();
+    unsigned idx = resAndIdx.index();
+    if (res.isFormalDirect()) {
+      if (auto dirRes = usedDirectResults[dirResIdx])
+        if (dirRes && activityInfo.isUseful(dirRes, parentIndices.source))
+          resultIndices.push_back(idx);
+      ++dirResIdx;
+    } else {
+      if (activityInfo.isUseful(arguments[indResIdx].get(),
+                                parentIndices.source))
+        resultIndices.push_back(idx);
+      ++indResIdx;
+    }
+  }
+}
+
+LinearMapInfo::LinearMapInfo(ADContext &context,
+                             AutoDiffAssociatedFunctionKind kind,
+                             SILFunction *original, SILFunction *assocFn,
+                             const SILAutoDiffIndices &indices,
+                             const DifferentiableActivityInfo &activityInfo,
+                             SILBuilder &builder)
+    : kind(kind), original(original), activityInfo(activityInfo),
+      typeConverter(context.getTypeConverter()), builder(builder) {
+  populateLinearMapStructDeclarationFields(context, indices, assocFn);
+}
+
+bool LinearMapInfo::shouldBeDifferentiated(ApplyInst *ai,
+                                           const SILAutoDiffIndices &indices) {
+  // Anything with an active result should be differentiated.
+  if (llvm::any_of(ai->getResults(), [&](SILValue val) {
+    return activityInfo.isActive(val, indices);
+  })) {
+    return true;
+  }
+  // Function applications with an active indirect result should be
+  // differentiated.
+  for (auto indRes : ai->getIndirectSILResults())
+    if (activityInfo.isActive(indRes, indices)) {
+      return true;
+    }
+  // Function applications with an inout argument should be differentiated.
+  auto paramInfos = ai->getSubstCalleeConv().getParameters();
+  for (auto i : swift::indices(paramInfos))
+    if (paramInfos[i].isIndirectInOut() &&
+        activityInfo.isActive(
+            ai->getArgumentsWithoutIndirectResults()[i], indices)) {
+      return true;
+    }
+  // Instructions that may write to memory and that have an active operand
+  // should be differentiated.
+  if (ai->mayWriteToMemory())
+    for (auto &op : ai->getAllOperands())
+      if (activityInfo.isActive(op.get(), indices)) {
+        return true;
+      }
+  return false;
+}
+
+ void LinearMapInfo::populateLinearMapStructDeclarationFields(
```

Comment:
I'm fine with checking this function to unblock things, but it should be broken down into multiple functions to reduce nesting.

Reply:
No worries I can make it part of this PR. Currently rebuilding my Xcode build, so will push a separate commit to make this change soon once it finishes.

Reply:
Added this change in my latest PR, separated out the code of when I find an apply instruction.

Reply:
Thanks!